sdk/core/Azure.Core/src/Pipeline/AzureOperationScope.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+
+namespace Azure.Core.Pipeline
+{
+    public struct DiagnosticScope: IDisposable
+    {
+        private Activity _activity;
+
+        private readonly string _name;
+
+        private readonly DiagnosticListener _source;
+
+        internal DiagnosticScope(string name, DiagnosticListener source)
+        {
+            _name = name;
+            _source = source;
+            _activity = _source.IsEnabled() ? new Activity(_name) : null;
+        }
+
+        public bool IsEnabled => _activity != null;
+
+        public void AddAttribute(string name, string value)
+        {
+            _activity?.AddTag(name, value);
+        }
+
+        public void AddAttribute<T>(string name, T value)
+        {
+            if (_activity != null)
+            {
+                AddAttribute(name, value.ToString());
+            }
+        }
+
+        public void AddAttribute<T>(string name, T value, Func<T, string> format)
+        {
+            if (_activity != null)
+            {
+                AddAttribute(name, format(value));
+            }
+        }
+
+        public void Start()
+        {
+            if (_activity != null && _source.IsEnabled(_name))
+            {
+                _source.StartActivity(_activity, null);
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_activity == null)
+            {
+                return;
+            }
+
+            if (_source != null)
+            {
+                _source.StopActivity(_activity, null);
+            }
+            else
+            {
+                _activity?.Stop();
+            }
+        }
+
+        public void Failed(Exception e)
+        {
+            if (_activity == null)
+            {
+                return;
+            }
+
+            _source?.Write(_activity.OperationName + ".Exception", e);
```

Comment:
Most likely it will be ignored by tracing system as it is likely handled and just noise. Unhandled exceptions are taken care of by web framework, So I think we don't need it for now.

Reply:
Don't we care about setting a status flag on spans?

Reply:
we do, can we do it better than Exception instance? If not, that's ok, but we need a plan

Reply:
We can, but using an exception should be an option too.